classes/boardmanager.php

```diff
@@ -0,0 +1,1152 @@
+     * Get a discussion record.
+     *
+     * @param int $messageid Id of the message
+     * @return stdClass
+     */
+    public function get_discussion_message(int $messageid): stdClass {
+        global $DB;
+        return $DB->get_record('kanban_discussion_comment', ['id' => $messageid], '*', MUST_EXIST);
+    }
+
+    /**
+     * Get cm_info object to current instance.
+     *
+     * @return cm_info
+     */
+    public function get_cminfo(): cm_info {
+        return $this->cminfo;
+    }
+
+    /**
+     * Writes a record to the history table.
+     *
+     * @param string $action Action for history
+     * @param int $type Type of object affected by the entry
+     * @param array $data Array of data to write
+     * @param int $columnid Id of the column
+     * @param int $cardid Id of the card
+     */
+    public function write_history(string $action, int $type, array $data = [], int $columnid = 0, int $cardid = 0): void {
+        global $DB, $USER;
+
+        if (empty($this->kanban->history) || empty(get_config('mod_kanban', 'enablehistory'))) {
+            return;
+        }
+
+        $affecteduser = null;
+        // Affected user must be written to a separate column (for privacy provider).
+        if (!empty($data['userid'])) {
+            $affecteduser = $data['userid'];
+            unset($data['userid']);
+        }
+        // Unset unused data.
+        unset($data['timemodified']);
+        unset($data['timecreated']);
+        unset($data['createdby']);
+        unset($data['canedit']);
+        unset($data['id']);
+        $record = [
+            'action' => $action,
+            'kanban_board' => $this->board->id,
+            'userid' => $USER->id,
+            'kanban_column' => $columnid,
+            'kanban_card' => $cardid,
+            'parameters' => json_encode($data),
+            'affected_userid' => $affecteduser,
+            'timestamp' => time(),
+            'type' => $type,
+        ];
+        $DB->insert_record('kanban_history', $record);
+    }
+
+    /**
+     * Update completion state
+     *
+     * @param array $users Array of userids or user records (if empty, current user is used)
+     * @return void
+     */
+    public function update_completion(array $users = []): void {
+        global $USER;
+        if (empty($users)) {
+            $users = [$USER->id];
+        }
+        if ($this->completion_enabled()) {
+            $completion = new \completion_info($this->course);
+            foreach ($users as $user) {
+                if (is_object($user)) {
+                    $completion->update_state($this->cminfo, COMPLETION_UNKNOWN, $user->id);
+                } else {
+                    $completion->update_state($this->cminfo, COMPLETION_UNKNOWN, $user);
+                }
+            }
+        }
+    }
+
+    /**
+     * Whether the custom completion rules are enabled for this board.
+     *
+     * @return bool
+     */
+    public function completion_enabled(): bool {
```

Comment:
Maybe this should be named `custom_completion_enabled` instead to be superclear?